src/__private_api.rs

```diff
@@ -0,0 +1,321 @@
+
+mod sealed {
+    use crate::Level;
+    use std::fmt::Arguments;
+
+    pub trait LogLevel {
+        fn into_log_level(self) -> Level;
+    }
+
+    pub trait LogArgs {
+        fn with(self, f: impl FnOnce(Arguments));
+    }
+
+    pub trait LogTarget {
+        fn with(self, module_path: &'static str, f: impl FnOnce(&str));
+    }
+
+    pub trait LogKVs {
+        fn with(self, f: impl FnOnce(&[(&str, &super::LogKvValue)]));
+    }
+}
+
+// `LogLevel`.
+
+impl LogLevel for Level {
+    #[inline]
+    fn into_log_level(self) -> Level {
+        self
+    }
+}
+
+macro_rules! define_static_levels {
+    ($($ty:ident => $lvl:ident,)*) => {
+        $(
+            #[derive(Debug)]
+            pub struct $ty;
+
+            impl LogLevel for $ty {
+                #[inline]
+                fn into_log_level(self) -> Level {
+                    Level::$lvl
+                }
+            }
+
+            impl PartialEq<LevelFilter> for $ty {
+                #[inline]
+                fn eq(&self, other: &LevelFilter) -> bool {
+                    Level::$lvl.eq(other)
+                }
+            }
+
+            impl PartialOrd<LevelFilter> for $ty {
+                #[inline]
+                fn partial_cmp(&self, other: &LevelFilter) -> Option<Ordering> {
+                    Level::$lvl.partial_cmp(other)
+                }
+
+                #[inline]
+                fn lt(&self, other: &LevelFilter) -> bool {
+                    Level::$lvl.lt(other)
+                }
+
+                #[inline]
+                fn le(&self, other: &LevelFilter) -> bool {
+                    Level::$lvl.le(other)
+                }
+
+                #[inline]
+                fn gt(&self, other: &LevelFilter) -> bool {
+                    Level::$lvl.gt(other)
+                }
+
+                #[inline]
+                fn ge(&self, other: &LevelFilter) -> bool {
+                    Level::$lvl.ge(other)
+                }
+            }
+        )*
+    };
+}
+
+define_static_levels![
+    StaticLevelError => Error,
+    StaticLevelWarn => Warn,
+    StaticLevelInfo => Info,
+    StaticLevelDebug => Debug,
+    StaticLevelTrace => Trace,
+];
+
+// `LogArgs`.
+
+impl LogArgs for &str {
+    #[inline]
+    fn with(self, f: impl FnOnce(Arguments)) {
+        f(format_args!("{self}"))
+    }
+}
+
+impl LogArgs for Arguments<'_> {
+    fn with(self, f: impl FnOnce(Arguments)) {
+        f(self)
+    }
+}
+
+// `LogTarget`.
+
+impl LogTarget for &str {
+    #[inline]
+    fn with(self, _module_path: &'static str, f: impl FnOnce(&str)) {
+        f(self)
+    }
+}
+
+#[derive(Debug)]
+pub struct TargetIsModulePath;
+
+impl LogTarget for TargetIsModulePath {
+    #[inline]
+    fn with(self, module_path: &'static str, f: impl FnOnce(&str)) {
+        f(module_path)
+    }
+}
+
+// `LogKVs`.
+
+impl LogKVs for &[(&str, &LogKvValue<'_>)] {
+    #[inline]
+    fn with(self, f: impl FnOnce(&[(&str, &LogKvValue)])) {
```

Comment:
Why every `Log*` trait has to accept a closure here?
Can't they just return the value directly?

Returning the closure would only add more overhead and generate more code.

Reply:
For `LogArgs`, I can’t return a `Arguments` value from a `&str` value:

```rust
fn f(s: &str) -> Arguments<'_> {
    format_args!("{s}") // Does not work because a local variable reference is captured.
}
```

So I have to use a callback function.

For other ones, since they need to be converted into references, the trait method may need to take a `&self` argument, and I don’t feed like taking references from `&str` or a zero sized type, since the will get a `&&str` type and `&ZST` type. But I am not very insistent on this.

Reply:
> So I have to use a callback function.

Yeah for `LogArgs` you have to use a callback due to limitation of `Arguments<'_>`, although I wonder
 
> For other ones, since they need to be converted into references, the trait method may need to take a `&self` argument, and I don’t feed like taking references from `&str` or a zero sized type, since the will get a `&&str` type and `&ZST` type. But I am not very insistent on this.

I don't think there's much issue with this since it will most likely be inlined.

Reply:
Now I have removed callback functions from `LogTarget` and `LogKVs`.